### ISSUE TITLE
SessionMiddleware can not work on swoole

### ISSUE THREAD

**Issue by ChisWill** [yiisoft/session#25]
Hi guys，my application run with swoole, I found I can not use SessionMiddleware because Session will be overwrite by other request.
Could we add Session into attribute of request, so we can use the way `$request->getAttribute(SessionInterface::class)` to get Session.
We also can remain old usage by add a new property when application running different env.

**Comment by samdark** [comment]
Do you want to keep session in the worker context and not to destroy/recreate it on each request?

**Comment by ChisWill** [comment]
No, I will create a new PsrRequest instance on each request and this is not my point.
I mean, when we use SessionInterface as a singleton and application is running on swoole，the code 
```php
if ($requestSessionId !== null && $this->session->getId() === null) {
    $this->session->setId($requestSessionId);
}
```
will modify session singleton and never execute `setId()` again when new request come.
The result is every request has same session.

**Comment by samdark** [comment]
Wait, but `$requestSessionId` is obtained from request and you have different request object every time, don't you?

```php
$requestSessionId = $this->getSessionIdFromRequest($request);
```

**Comment by ChisWill** [comment]
Yes, request is different.
But, look at the if condition `if ($requestSessionId !== null && $this->session->getId() === null) {` 
`$this->session` had been modified and storage in memory, so after first request the `$this->session->getId()` is always not null.
And at the last  SessionMiddleware will `commitSession()`, it will modify current request session id to a same value which is `$this->session->getId()`.

**Comment by samdark** [comment]
You're right. Any fix that comes in mind? I don't currently have Swoole environment set up :(

**Comment by ChisWill** [comment]
As I mentioned in the question, we should use the code `$request->getAttribute(SessionInterface::class)` to get Session.
I think Session depends on request, it shouldn't be a singleton instance.
However, this will affect the usage of the current user, so add a property to configure?
It's up to you.

**Comment by yiiliveext** [comment]
> You're right. Any fix that comes in mind? I don't currently have Swoole environment set up :(

Make `sessionId` nullable.
```php
public function setId(?string $sessionId): void
{
    $this->sessionId = $sessionId;
}
```
And we should add `resetter` in the `SessionInterface::class` config for those who use `yiisoft/di`

**Comment by ChisWill** [comment]
I think it won't work well.
In Swoole coroutine environment，suppose this situation:
Even if there is only one process and the first request come and do `sleep()`, the process will also accept other request.
At this time, the other request will modify the session, Until the sleeping request wakes up, his session had been changed.
So I think every request should maintain its own session copy.

**Comment by samdark** [comment]
1. Added resetter config.
2. Updated https://github.com/yiisoft/docs/blob/master/guide/en/tutorial/using-yii-with-swoole.md

@ChisWill ping us if it doesn't help please.

**Comment by ChisWill** [comment]
> 1. Added resetter config.
> 2. Updated https://github.com/yiisoft/docs/blob/master/guide/en/tutorial/using-yii-with-swoole.md
> 
> @ChisWill ping us if it doesn't help please.

I refer to this article, but I found that in the same process  `session_id()` always returns the same value when a new request comes.
Why not use `session_create_id()` or  other similar method?

**Comment by alamagus** [comment]
@ChisWill 
Php cli(on top of which swoole is run)  has no idea about coroutines. So, at any given time there is 0 or 1 session in the process. And when you call session_* functions 'simultaneously' from different coroutines, you still operate on 1(or 0) session
[src](https://github.com/php/php-src/blob/master/ext/session/session.c)
Might be wrong tho :thinking:

**Comment by klsoft-web** [comment]
I can confirm that Swoole overwrites the session state with that of another request.
In theory, we could use the [symfony/lock](https://packagist.org/packages/symfony/lock) package to guarantee exclusive access to the session state. 
However, this would eliminate all the performance benefits of Swoole and require a solution that goes beyond the scope of the `yiisoft/session` package.
In the Swoole repository, I found a [comment](https://github.com/swoole/swoole-src/issues/4478#issuecomment-962840796) advising against using PHP sessions with Swoole.

**Comment by samdark** [comment]
I think we need to add this info to https://github.com/yiisoft/docs/blob/master/src/guide/tutorial/using-yii-with-swoole.md and close the issue. Adding locks (we have yiisoft/mutex btw.) isn't a great idea in this case.

**Comment by klsoft-web** [comment]
I was wrong in my previous conclusions.
This is what I have recently discovered.

If you are using Swoole, set the 'worker_num' parameter to 1.
```php
$server = new Swoole\Http\Server('127.0.0.1', 9501);
$server->set(['worker_num' => 1]);
```

All requests with an empty `PHPSESSID` cookie, as well as all new requests, will receive the same session ID.
Calling the `open(`) method of `Yiisoft\Session\Session` will then [create that session ID](https://github.com/yiisoft/session/blob/master/src/Session.php#L102).
This issue can be resolved by replacing [this part of the code](https://github.com/yiisoft/session/blob/master/src/SessionMiddleware.php#L27C1-L29C10):
```php
if ($requestSessionId !== null && $this->session->getId() === null) {
    $this->session->setId($requestSessionId);
}
```

with the following:
```php
if ($this->session->getId() === null) {
    $this->session->setId($requestSessionId !== null ? $requestSessionId : session_create_id());
}
```

Software versions:
Swoole 6.2.2
yiisoft/session 3.0.2

**Comment by samdark** [comment]
Running it with just a single worker doesn't make sense.

**Comment by klsoft-web** [comment]
> Running it with just a single worker doesn't make sense.

Yes, using a single worker wouldn't make sense.
However, the number of workers doesn't matter for this test.
Here's how to reproduce the bug with more than one worker.
Set the 'worker_num' parameter to 2.
```php
$server = new Swoole\Http\Server('127.0.0.1', 9501);
$server->set(['worker_num' => 2]);
```
Then lock the first worker:
```php
$queryParams = $request->getQueryParams();
if (isset($queryParams['w'])) {
 sleep(60*10);
}
```

All subsequent requests will then be handled by the second worker.
In the end, we got the same result!
All requests with an empty PHPSESSID cookie, as well as all new requests, will receive the same session ID.

So, I think the issue occurred when any requests made after the first request with an empty PHPSESSID cookie were handled by the same worker.

**Comment by klsoft-web** [comment]
I repeated the above test with the [yiisoft/yii-runner-roadrunner](https://github.com/yiisoft/yii-runner-roadrunner) package and encountered the same issue.
 With:
```yml
   pool:
    num_workers: 1
```
    
 Or with:
```yml
   pool:
    num_workers: 2
```
and locking the first worker:
```php
$queryParams = $request->getQueryParams();
if (isset($queryParams['w'])) {
 sleep(60*10);
}
```

All requests with an empty PHPSESSID cookie, received the same session ID.
